src/index.ts

```diff
@@ -1,26 +1,144 @@
+import moment from 'moment';
+import { Request, Response } from 'express';
 import fs from "fs";
 import { cwd } from "process";
 import { Environments } from "./enums/environments";
+import { get, postWithConfig } from './utils/request';
 
 const DEFAULT_CONFIG_FILE_LOCATION = `${cwd()}/.bluebutton-config.json`;
 const SANDBOX_BASE_URL = "https://sandbox.bluebutton.cms.gov";
 const PRODUCTION_BASE_URL = "https://api.bluebutton.cms.gov";
+const GENERAL_DATA_ERR = '{"message": "Unable to load data - query FHIR resource error."}';
 
+type CodeChallenge = {
+  codeChallenge: string,
+  verifier: string
+};
+  
 type BlueButtonJsonConfig = {
   clientId: string;
   clientSecret: string;
   callbackUrl: string;
   version?: string;
+  pkce?: boolean;
+  retryConfig?: RetryConfig;
   environment?: Environments;
 };
 
+export type RetryConfig = {
+  enabled: boolean;
+  initInterval: number;
+  maxAttempts: number;
+  backOffExpr: string;
+  endpointPattern: string;
+};
+  
+export interface IAuthorizationToken {
+  accessToken: string,
+  expiresIn: number,
+  tokenType: string,
+  scope: [string],
+  refreshToken: string,
+  patient: string,
+  expiresAt: number
+}
+
+export class AuthorizationToken implements IAuthorizationToken {
+  public accessToken: string;
+  public expiresIn: number;
+  public expiresAt: number;
+  public tokenType: string;
+  public scope: [string];
+  public refreshToken: string;
+  public patient: string;
+  
+  constructor(authToken: any) {
+    this.accessToken = authToken.access_token;
+    this.expiresIn = authToken.expires_in;
+    this.expiresAt = authToken.expires_at ? authToken.expires_at : moment().add(this.expiresIn).valueOf();
+    this.patient = authToken.patient;
+    this.refreshToken = authToken.refresh_token;
+    this.scope = authToken.scope;
+    this.tokenType = authToken.token_type;
+  }
+}
+    
+export interface IUser {
```

Comment:
We should have no reference to a user here. That is up to the 3rd party app to implement, not the SDK.